### ISSUE TITLE
Cloudflare APO Cache Purge not working on the latest version

### ISSUE THREAD

**Issue by nsgoyat** [cloudflare/Cloudflare-WordPress#456]
### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the plugin and it is still present.

### WordPress version

5.9.2

### Cloudflare-WordPress version

4.8.0

### PHP version

8.1

### Expected result

The cache should be purged from the Cloudflare servers when we update the site content.

### Actual result

While using the latest version of the plugin and APO, if we update the site content, the Cloudflare Cache isn't purged and it still shows the old content.

### Steps to reproduce

1. Setup the latest Cloudflare Plugin on WordPress.
2. Enable APO settings.
3. Add a Page or article and get it cached on the Cloudflare Servers.
4. Update the article content from the WordPress dashboard and then open the URL in a private window of your browser.
5. You will notice that cache isn't purged from the Cloudflare and it still serving the old content.

### Additional factoids

APO Cache Purging works fine when we switch the plugin to version 4.7.0

**Comment by jacobbednarz** [comment]
What does https://github.com/cloudflare/Cloudflare-WordPress/blob/master/src/WordPress/Hooks.php#L351 evaluate yo for you? Does it show as APO enabled?

**Comment by jacobbednarz** [comment]
Similarly with https://github.com/cloudflare/Cloudflare-WordPress/blob/master/src/WordPress/Hooks.php#L365.

**Comment by nsgoyat** [comment]
Hey Jacob,
Unfortunately, I am not experienced enough to understand or perform the tasks you have listed above. May you please shed some more details on the debugging process?

**Comment by shashikant-rohit** [comment]
Hey Jacob,

Sorry to post this here but unfortunately this issue isn't fixed yet with the new release. Specially for Woocommerce Products. The products are still showing stale prices until refreshed even after updating from the WP Dashboard.

Could you please help us if anything else has to be done along with updating plugin from 4.8.1?

Ref Community Link : [](https://community.cloudflare.com/t/woocommerce-product-prices-not-changing-properly/369896/21)

**Comment by jacobbednarz** [comment]
What are the URLs in question that you are having issues with? 
Do you have any debug logs to share? Are the covered by the hook defined at https://github.com/cloudflare/Cloudflare-WordPress/blob/master/src/WordPress/Hooks.php#L218?

**Comment by shashikant-rohit** [comment]
Hi Jacob,

I am not a technical guy to understand the hooks or code. Bascically we are trying to change the prices of the products from WP Dashboard & when we check on the front end as an admin user - the prices change with immediate effect as the admin users are not covered under cache (bypassed it). But as a user when we try in a separate browser we will only see the old price in the Category Listing Page & when we click on the product & it opens up then the prices are correct.

When the listings page is refreshed also the price changes but not auto purged.

Below is the URL for your reference. The first product which you see - the price should be Rs. 17020 (updated 5min back) but it shows Rs. 17030 (old price)

[https://www.pcstudio.in/product-category/motherboard/](https://www.pcstudio.in/product-category/motherboard/)

**Comment by jacobbednarz** [comment]
Do you have APO enabled? Or any page rules that override the cache behaviour to cache everything?

**Comment by shashikant-rohit** [comment]
We have APO Enabled with wordpress plugin installed. 

Also we have few page rules as below. Few are for security & a couple for caching

By the way - with the same setup we have been running the website since almost 8 months now but only after 15th March this issue has started troubling.

![image](https://user-images.githubusercontent.com/64070861/158951457-ed816fff-4a49-466e-bba8-06f5211de11a.png)

**Comment by jacobbednarz** [comment]
thanks, i've found your requests internally and it looks like they are making it out of WordPress and into our systems. i'll chat with the cache folks about this and see what the potential issue is. by the plugin logic, it should still be purging everything (not restricting to extensions).

what is in your WordPress debug logs? it should show some HTTP requests being made for the purge calls.

**Comment by jacobbednarz** [comment]
for reference, there are definitely purges happening. these have been in the last 20 minutes.

```
www.pcstudio.in/product-category/brand/brand-msi/
www.pcstudio.in/brand/msi/feed/
www.pcstudio.in/motherboard-cpu-support/intel-lga-1700/feed/
www.pcstudio.in/product-category/motherboard/feed/          
www.pcstudio.in/product-category/motherboard/feed/          
www.pcstudio.in/author/vikas/                               
www.pcstudio.in/feed/                                       
www.pcstudio.in/product-category/brand/brand-msi/feed/      
www.pcstudio.in/product-category/motherboard/     
```

**Comment by shashikant-rohit** [comment]
As I can see currently the Debug.log file is empty & not been used recently from the File Manager of hosting server.

Also, you mean that there is some issue & the Cache team is working on it?

If purges are happening & then also the content updates aren't happening there seems to be some issue with the caching system or plugin.

**Comment by jacobbednarz** [comment]
no, there is no issue as far as i or the team can see. we're getting the purge requests (as above). are you still running v4.8.1? your debug logs should be showing the HTTP requests that are being made above.

**Comment by shashikant-rohit** [comment]
As of now there is no debug log file which contains entries but the php error log shows below entries after enabling the debug mode.

[18-Mar-2022 07:12:52 UTC] PHP Notice:  wp_woocommerce_session_a5d241075b0435145ca346318a1d2d9a cookie cannot be set - headers already sent by /home/pcstudio/public_html/wp-cron.php on line 28 in /home/pcstudio/public_html/wp-content/plugins/woocommerce/includes/wc-core-functions.php on line 1059

[18-Mar-2022 07:12:52 UTC] PHP Notice:  ob_end_flush(): failed to send buffer of zlib output compression (0) in /home/pcstudio/public_html/wp-includes/functions.php on line 5212

**Comment by jacobbednarz** [comment]
the version I am seeing internally is v4.7.0. so either you’ve upgraded/downgraded or the issue isn’t related to the upgrade.

**Comment by shashikant-rohit** [comment]
Earlier it was 4.8.0 and as per the discussion on cloudflare community I have tried downgrading the version to 4.7.0 which also didn't work.

This morning when I have seen that the issue was fixed in the latest version of plugin 4.8.1 I have again upgraded that plugin.

Are you saying the website is still on 4.7.0?

**Comment by jacobbednarz** [comment]
I am seeing v4.7.0 coming through in the last 30 minutes. Can you please upgrade to v4.8.1 and try updating posts again? Checking debug logs along the way for the outgoing HTTP requests?

**Comment by shashikant-rohit** [comment]
That a prob. Currently the version installed on the dashboard is 4.8.1. Below is the screen shot for your reference.

I have tried updating the posts after the update only & then sent you the URL

I have again changed the price of same product back to original. Can you check if any logs are being made or requests for caching?

![image](https://user-images.githubusercontent.com/64070861/158955446-e43788cf-6290-497f-8761-8946955f3ac9.png)

**Comment by jacobbednarz** [comment]
these are the last logs i can see making it to us (UTC). ~ 15m ago

```
2022-03-18 07:12:11 www.pcstudio.in/brand/msi/feed/
2022-03-18 07:12:11 www.pcstudio.in/brand/msi/     
2022-03-18 07:12:10 www.pcstudio.in/feed/          
2022-03-18 07:12:10 www.pcstudio.in/feed/atom/     
2022-03-18 07:12:10 www.pcstudio.in/brand/msi/feed/
2022-03-18 07:12:10 www.pcstudio.in/               
2022-03-18 07:12:10 www.pcstudio.in/brand/msi/     
2022-03-18 07:12:10 www.pcstudio.in/feed/atom/     
2022-03-18 07:12:10 www.pcstudio.in/brand/msi/feed/
2022-03-18 07:12:10 www.pcstudio.in/brand/msi/feed/
```

i'm beginning to think this is a separate issue 🤔

**Comment by shashikant-rohit** [comment]
OMG !

Seems to be some other issue then. Do I need to raise a separate thread here on github?

Let me know the way forward.

Guess only after the last update on March 15th this started.

Also, these requests which are coming to you are from 4.7.0 or 4.8.1? Can you identify that?

**Comment by jacobbednarz** [comment]
the user agent does look wonky as it is sending v4.7.0 (i'll look into that separately). your pages are definitely getting purged (note the `age` being so low).

```

~ ❯ curl -svo /dev/null https://www.pcstudio.in/product-category/motherboard/ 

* Connection state changed (MAX_CONCURRENT_STREAMS == 256)!
< HTTP/2 200 
< date: Fri, 18 Mar 2022 07:33:54 GMT
< content-type: text/html; charset=UTF-8
< cf-ray: 6edc4867aaf50c95-EWR
< age: 2
< cache-control: max-age=86400
< last-modified: Fri, 18 Mar 2022 07:16:29 GMT
< link: <https://www.pcstudio.in/wp-json/>; rel="https://api.w.org/", <https://www.pcstudio.in/wp-json/wp/v2/product_cat/18>; rel="alternate"; type="application/json"
< vary: Accept-Encoding
< cf-cache-status: HIT
< cf-apo-via: tcache
< cf-edge-cache: cache,platform=wordpress
< expect-ct: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
< x-ua-compatible: IE=edge
< report-to: {"endpoints":[{"url":"https:\/\/[a.nel.cloudflare.com](http://a.nel.cloudflare.com/)\/report\/v3?s=lKe2hRx%2F%2BNS00t5Z0SxJGVZITn0kOvaQxIHqbW0TUf4QAq2Z9cSfPdxbunXGX7ManyKIjj3wzjJzfK2gMDNgQi7eD6A1PseTycxMzy%2BSe8uXC8b8T6Jx9PeWc92DBM3IzA%3D%3D"}],"group":"cf-nel","max_age":604800}
< nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
< server: cloudflare
< 
{ [671 bytes data]
* Connection #0 to host [www.pcstudio.in](http://www.pcstudio.in/) left intact

```

i would suggest you open a support ticket with Cloudflare including the details from here and that you are having issues on 4.7.0 _and_ 4.8.x to ensure we don't go through this all over again. you can also ask the support team to pass it to the APO team who will be able to dig into this a little more too.

**Comment by shashikant-rohit** [comment]
Thanks Jacob. I have raised a support ticket from our dashboard with a detailed summary on what's the issue and series of events which happened from last couple of days with the link of this thread as well as the community thread link.

Hoping to hear back from them soon.

We are facing this issue from last couple of days & this gives users a bad experiencen.

Let me know if you have any update on the issue which you wanted to check separately (requests coming from 4.7.0 even after update to 4.8.1)

Thanks Again
Shashikant

**Comment by jacobbednarz** [comment]
do you have a ticket ID? i can give the team a heads up to keep an eye on it.

**Comment by firatcf** [comment]
Hi folks,

We released an update for our plugin. Can you please download it and give it a try?

Please let me know if you still see the issue.

Thank you.

**Comment by jacobbednarz** [comment]
just FYI, v4.8.2 is a no op functionality wise. I only fixed the user agent to show the current version in use.

**Comment by shashikant-rohit** [comment]
Hi @firatcf ,

I don't think this update is working.

We have just updated the version to the latest one 4.8.2 & tried changing the prices but still it shows the old prices.

Thanks
Shashikant

**Comment by shashikant-rohit** [comment]
Hey,

The issue seems to be fixed now I guess. The prices are getting updated now. Did you work on something or update anything on the plugin side?

Also, can you please let me know usually how much time it takes for the posts to be updated on the front end when caching is used?

Will recheck on couple of other updates & confirm you back if the issue is fixed completely


Thanks
Shashikant

**Comment by firatcf** [comment]
Hi @shashikant-rohit ,

That's good to hear. Please let us know if you experience this issue again.

It should be within 30 seconds.

**Comment by shashikant-rohit** [comment]
Hi @firatcf

Some time this is working & sometime not.

I have just updated few product prices and waited for alomost 25 minutes & still the product are showing old prices.

Seems like this isn't fixed temporarily or fully.

Could you please help know what's the issue and how can this be fixed permanently?


Thanks
Shashikant

**Comment by jacobbednarz** [comment]
@shashikant-rohit let’s take this to the support ticket. it’s no longer the issue mentioned here and it’s best served through our usual support methods. from a plug-in perspective, this is working as intended.
